### ISSUE TITLE
As a developer, it would be awesome if this DDEV plugin set up my phpunit for me

### ISSUE THREAD

**Issue by cosmicdreams** [php-perfect/ddev-intellij-plugin#166]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you sure that this bug is related to this DDEV Integration Plugin?

- [X] I am sure

### Describe the bug

When using DDEV in PhpStorm there is a bug that is caused by how the tool decides to set the path mapping for a DDEV-driven project.  PhpStorm appears to discover that there is a path mapped to `/var/tmp/html` when mutagen is enabled.  The expected directory to map for the project root is `/var/www/html`.  Through debugging the issue with DDEV developers, we discovered the workaround is easy, just manually set the path mapping for project root to be `/var/www/html`, manually.

I'm hoping the plugin can help make this workaround be a part of the help and assistance that is provided.  I've been thinking about the plugin as the thing that makes using DDEV with PhpStorm easy.  Fixing a flaw that is seemingly caused by how PhpStorm decides to create these mappings would be a great thing to "patch".

Related DDEV troubleshooting: https://github.com/ddev/ddev/issues/4697
Fix to DDEV Documentation: https://github.com/ddev/ddev/pull/4732

### Steps to reproduce

Setup a Drupal site using a modified version of the Drupal 10 quickstart.  This adds in Selenium for functional testing and phpstorm dependencies.

1. cd ~/Sites
2. mkdir my-drupal10-site
3. cd my-drupal10-site
4. ddev config --project-type=drupal10 --docroot=web --create-docroot
5. ddev get ddev/ddev-selenium-standalone-chrome 
6. ddev start
7. ddev composer create drupal/recommended-project
8. ddev composer require drupal/core-dev --dev --update-with-all-dependencies
8. ddev composer require drush/drush
9. ddev drush site:install --account-name=admin --account-pass=admin -y

## Add Mutagen
- `ddev config --mutagen-enabled=true`
- `ddev start`

## Testing the bug
1. Open the project in PhpStorm
2. Open preferences and navigate to PHP to view the interpreter that this plugin setup.
3. Click the folder icon in the `Path Mappings` for the interpreter 
4. Observe that the project root is mapped to `/var/tmp/html`

**Comment by rfay** [comment]
I'm not sure this is in scope for the plugin, but if it is, always making the default project root mapping be Project Root -> /var/www/html would make everything work always.

**Comment by rfay** [comment]
@cosmicdreams please change the title; Maybe "Wrong project root mapping provided by default set when Mutagen enabled"

**Comment by cosmicdreams** [comment]
title fixed.

**Comment by nico-loeber** [comment]
Hey @cosmicdreams,

Thanks for reporting this! 
If there is something the plugin can do to avoid manual configuration steps, this is something we should definitely look into further!
Could you please show a screenshot of the wrong mapping path? Currently the project root inside the container is set statically, as it is always the same for ddev containers, see: 

https://github.com/php-perfect/ddev-intellij-plugin/blob/c83e5d0ab467e09d767ce6a221cc28c4756dacce/src/main/java/de/php_perfect/intellij/ddev/php/server/ConfigureServerListener.java#L39

Thanks in advance!

**Comment by cosmicdreams** [comment]
Here's a screenshot of what my "fix" had to be.  It shows how I manually set the path to project root to /var/www/html AND how the system set the project root to /var/tmp/html

![Screenshot 2023-03-09 at 9 27 10 AM](https://user-images.githubusercontent.com/852080/226118897-138fb057-fb63-47b1-acf9-2c49a8839ad3.png)

Here's a screenshot of the config that I believe led the system into thinking that was the right thing to do

![Screenshot 2023-03-18 at 11 22 30 AM](https://user-images.githubusercontent.com/852080/226119012-64263177-02d7-43c9-9148-f05eef896d3f.png)

You see here that it's setting a path binding to /var/tmp/html

Desired fix:
- Within the plugin's logic.
- Check if mutagen is enabled
- Check what the project's path mapping for the project root is (within PhpStorm's path mapping config)
- if it's set to /var/tmp/html, override it to be /var/www/html. 

The tmp directory is supposed to be read only, and shouldn't be the point of interaction for phpstorm because running tests sometimes means creating folders, files, screenshots, etc.  All of which need write access.

**Comment by rfay** [comment]
Mappings for all DDEV projects should be Project Root -> /var/www/html

**Comment by cosmicdreams** [comment]
@rfay the line of code that @nico-loeber showed seems to be doing that, and yet the bug occurred.  I wonder if there is an order of operations issue here.  Does the plugin's attempt to set the path mapping occur after something else that sets the path based upon the discovered bindings?

**Comment by rfay** [comment]
AFAIK this plugin doesn't manage mappings for phpunit.

**Comment by cosmicdreams** [comment]
Oh that too.  The phpunit configuration could be managed as the settings are the same each time, well for Drupal configs that is.  Here's a screenshot and let me break it down what the config normally looks like:

<img width="1266" alt="Screenshot 2023-03-18 at 1 26 03 PM" src="https://user-images.githubusercontent.com/852080/226127540-f1a52438-a9d8-4c85-9a59-b12cae19e7ef.png">

1. Get rid of the local interpreter settings that are normally associated with phpunit
2. Set up a remote interpreter to use the one that is already configured by the plugin.
3. Use PHAR method of finding phpunit
4. Set the path to the command to `/var/www/html/vendor/bin/phpunit`
5. Finally, for me (because I'm using the [ddev-selenium-standalone-chrome](https://github.com/ddev/ddev-selenium-standalone-chrome) library, I can just use the `/var/www/html/web/core/phpunit.xml.dist` default config file.  Maybe that would be fine for most scenarios.  Something to talk through.

**Comment by cosmicdreams** [comment]
Having the plugin properly setup EVERYTHING you need to get tests running would be great, but feels like a separate issue from what the purpose of this issue is about.

**Comment by cosmicdreams** [comment]
I think it's really cool that this plugin sets up your database connection, that's a huge "selling point" for when I advocate for DDEV and this plugin as core tools for our local development.  ALSO being able to have your test runner setup when a DDEV project using Drupal is discovered would be awesome.

**Comment by rfay** [comment]
> Having the plugin properly setup EVERYTHING you need to get tests running would be great, but feels like a separate issue from what the purpose of this issue is about.

There are other mappings that could be set up, for example, phpcs. But AFAICT this issue should be retitled and the summary fixed, as I think it is *exactly* about running phpunit. It's not about the regular mapping, which I think is done correctly.

That's also the same thing that https://github.com/ddev/ddev/pull/4732 is waiting for you on.

**Comment by nico-loeber** [comment]
@cosmicdreams, everything that is required to run PHPUnit, in PHPStorm should already be configured by the plugin.

* clone https://github.com/php-perfect/ddev-integration-demo in PHPStorm
* Click ddev start in phpstorm
* open WorldTest.php and click one of the play buttons
* the test suit should be run and succeed

If this does not work in the example project please let me know.

**Comment by cosmicdreams** [comment]
As I mentioned in the Discord channel, I am waiting for my colleague to independently retest the error to see if they encounter it.  But I have not been able to get the error to recur myself.

I've come around to the idea that this issue needs to be retitled and focus instead on having the plugin set up configuration for phpunit.

**Comment by cosmicdreams** [comment]
@nico-loeber Are we aligned on whether this plugin should establish phpunit config during it's initialization / discovery phase? Or whatever the phase is called when it's creating the interpreter config.

**Comment by nico-loeber** [comment]
Hi @cosmicdreams,

Thanks for summarizing!

Regarding the bug you were experiencing, please let me know if this happens to you or one of your colleagues again!

I am always willing to improve the plugin for a better integration of DDEV, this means to me, to reduce repetitive setup that is always the same for a DDEV environment, like configuring PHP Interpreters and database connections.
In my opinion the phpunit setup is none of them.
First of all you need to tell PHPStorm if you are using PHPUnit installed in your vendor directory or if you are using it as a phar installed by phive or maybe donwloaded manually. Both ways to setup phpunit are valid and they are not DDEV related. It is a project specific decission.
Also the location of your phpunit.xml might differ depending on your project structure. It is most common to store it in the project root of course.
In case you you have installed phpunit in your vendor using composer and your phpunit.xml(.dist) is located in your project root, PHPStorm will already detect it automatically and no PHPUnit configuration is required.
This is tested with the ddev integration demo project. If this is not working for you we can take a look at it, if you like. But please note this is a phpstorm feature and none of the plugin.

As there is no single "right" configuration for PHPUnit I would like to keep this out of the ddev setup process as this configuration would override the auto detection of PHPstorm that might give better results for people, sorry.

**Comment by cosmicdreams** [comment]
That would be excellent, we can cut down the communication loop and be able to point to configs from newly setup sites and stuff.

I will have time on Friday, April 15th to do such a screenshare.  What time / timezone would work best for you?

**Comment by rfay** [comment]
Note also that the internal mapping to /var/tmp/html is likely to go away in next DDEV version, so the confusing mapping might go away, but PhpStorm also might not be able to guess the correct mapping at all.

**Comment by nico-loeber** [comment]
> That would be excellent, we can cut down the communication loop and be able to point to configs from newly setup sites and stuff.
> 
> I will have time on Friday, April 15th to do such a screenshare. What time / timezone would work best for you?

Sorry, I'm a bit late...

I'm in UTC+2 (Europe/Berlin). Please write me at nico@php-perfect.de so we can find a date without sapmming this discussion, thanks!

**Comment by Gonzalo2683** [comment]
> I'm not sure this is in scope for the plugin, but if it is, always making the default project root mapping be Project Root -> /var/www/html would make everything work always.

Apologies for asking what might be a very basic question, but I'm setting up ddev, PhpStorm, Mutagen, and when I run the tests, I encounter problem:
```stacktrace
HP Warning:  file_put_contents(/var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/../../../../../../sites/simpletest/Assert.php): failed to open stream: Read-only file system in /var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/ClassWriter.php on line 115

Warning: file_put_contents(/var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/../../../../../../sites/simpletest/Assert.php): failed to open stream: Read-only file system in /var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/ClassWriter.php on line 115
PHP Warning:  include(/var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/../../../../../../sites/simpletest/Assert.php): failed to open stream: No such file or directory in /var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/ClassWriter.php on line 60

Warning: include(/var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/../../../../../../sites/simpletest/Assert.php): failed to open stream: No such file or directory in /var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/ClassWriter.php on line 60
PHP Warning:  include(): Failed opening '/var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/../../../../../../sites/simpletest/Assert.php' for inclusion (include_path='/var/tmp/html/vendor/pear/archive_tar:/var/tmp/html/vendor/pear/console_getopt:/var/tmp/html/vendor/pear/pear-core-minimal/src:/var/tmp/html/vendor/pear/pear_exception:.:/usr/share/php') in /var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/ClassWriter.php on line 60

Warning: include(): Failed opening '/var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/../../../../../../sites/simpletest/Assert.php' for inclusion (include_path='/var/tmp/html/vendor/pear/archive_tar:/var/tmp/html/vendor/pear/console_getopt:/var/tmp/html/vendor/pear/pear-core-minimal/src:/var/tmp/html/vendor/pear/pear_exception:.:/usr/share/php') in /var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/ClassWriter.php on line 60
PHP Warning:  file_put_contents(/var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/../../../../../../sites/simpletest/TestCase.php): failed to open stream: Read-only file system in /var/tmp/html/web/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/ClassWriter.php on line 115

```

From my research, I found that this could be due to the mapping being in /var/tmp/html, but I can't find where to manually change this in PhpStorm when using the ddev plugin for PhpStorm. Is there any documentation available on this integration that I can review?

**Comment by rfay** [comment]
@Gonzalo2683 you must remove the mapping that uses /var/tmp/html and use the mapping of Workspace->/var/www/html.

That's what this issue is about, trying to get the plugin to do this for you.

**Comment by nico-loeber** [comment]
Hey @cosmicdreams,

Finally I had some time to investigate this a bit further.

The problem was not related to the phpunit configuration at all. The source of the problem is that the manually set path Mapping when configuring the interpreter (Code referenced aboth) is ignored by PHPStorm if it detects the path mapping from the docker-compose file. This causes not only problems for phpunit but for all cli executions like composer installs.

Enabling mutagen changed the docker-compose file in a way that phpstorm can no longer find a path mapping for /var/www/html as @rfay already mentioned. The solution for this problem is setting the additional path mapping not on interpreter level but on project level in the configuration.
This is done in https://github.com/php-perfect/ddev-intellij-plugin/commit/c07efcb48ee4b691f282c441e35c17ab06fe3d1e.

This fix should be released as soon as I have finished another feature I am currently working on (Within the next few days).

Sorry for the waiting time. Feedback is welcome.

**Comment by nico-loeber** [comment]
The fix is relased in version 1.1.0 (pending JetBrains review). Please let me know if still have any issues, @cosmicdreams.

**Comment by shochdoerfer** [comment]
I just had a slight issue with getting PHPUnit to run via DDEV/Docker. It worked fine with the "Always start a new container" CLI lifecycle configuration but failed when I switched to "Connect to existing container". The error that popped up mentioned that the name did not match a regex. It turned out that the DDEV project name needs to be all lowercase and I had one uppercase character in it.
